tutorial/9 - expert.test.ts

```diff
@@ -0,0 +1,445 @@
+            );
+
+            /**
+             * ** Your Turn **
+             *
+             * `hasDerived` is used in the first derivation. But has it been
+             * called at this point?
+             */
+
+            // `.toHaveBeenCalled()` or `.not.toHaveBeenCalled()`? ↴
+            expect(hasDerived) /* Your Turn */;
+
+            mySecondDerivation$.get();
+
+            /**
+             * ** Your Turn **
+             *
+             * Now that we have gotten `mySecondDerivation$`, which calls
+             * `.get()` on the first multiple times. How many times has the
+             * first `Derivable` actually executed its derivation?
+             */
+            // how many times?
+            expect(hasDerived).toHaveBeenCalledTimes(__YOUR_TURN__);
+        });
+
+        /**
+         * So when a `Derivable` is reacting the value is cached and can be
+         * gotten from cache. But if this `Derivable` is used multiple times in
+         * a row, even in another derivation it isn't cached.
+         *
+         * To fix this issue, `.autoCache()` exists. It will cache the
+         * `Derivable`s value until the next Event Loop `tick`.
+         *
+         * So let's try the example above with this feature
+         */
+        it('autoCaching', async () => {
+            const firstHasDerived = jest.fn();
+            const secondHasDerived = jest.fn();
+
+            /**
+             * ** Your Turn **
+             *
+             * Use `.autoCache()` on one of the `Derivable`s below. To make the
+             * expectations pass.
+             */
+            const myAtom$ = atom(true);
+            const myFirstDerivation$ = myAtom$.derive(firstHasDerived);
+            const mySecondDerivation$ = myFirstDerivation$.derive(() =>
+                secondHasDerived(myFirstDerivation$.get() + myFirstDerivation$.get()),
+            );
+
+            // first before .get()
+            expect(firstHasDerived).not.toHaveBeenCalled();
+
+            // second before .get()
+            expect(secondHasDerived).not.toHaveBeenCalled();
+
+            mySecondDerivation$.get();
+
+            // first after first .get()
+            expect(firstHasDerived).toHaveBeenCalledOnce();
+            // second after first .get()
+            expect(secondHasDerived).toHaveBeenCalledOnce();
+
+            mySecondDerivation$.get();
+
+            // first after second .get()
+            expect(firstHasDerived).toHaveBeenCalledOnce();
+            // second after second .get()
+            expect(secondHasDerived).toHaveBeenCalledTimes(2);
+
+            /**
+             * Notice that the first `Derivable` has only been executed once,
+             * even though the second `Derivable` executed twice.
+             *
+             * Now we wait a tick for the cache to be invalidated.
+             */
+
+            await new Promise(r => setTimeout(r, 1));
+
```

Comment:
Hier werd in de oorspronkelijke versie de mock gecleared na de delay. Deze variant zonder de clear lijkt me echter iets interessanter omdat je dan het resultaat van die delay ook kan waarnemen.

Reply:
Het bracht mij wel even van mijn a propos... Moet even kijken wat het duidelijkst is.